### PR TITLE
Disable incompatible payment methods when manual capture is enabled

### LINKED ISSUE #9363: Payment methods do not declare support for Cart and Checkout blocks if they cannot be enabled in the Payments settings
### Describe the bug

If any payment methods cannot be enabled because...
1. Auth + capture is enabled
2. That method has been rejected
3. That method requires more information to enable

...the Cart and Checkout block will show those particular payment methods do not support those blocks.

### To Reproduce

1. Create a page that uses the Cart or Checkout block.
(if such a page already exists, edit that page) 
2. Select one of the blocks on the page.
3. Open the sidebar.
4. Select the `Block` tab.
5. Note there are no listed incompatibilities with any of the WooPayments payment methods listed in the sidebar.
6. Navigate to **Payments > Settings**.
7. Enable any BNPL payment method or LPM.
8. Enable `Issue an authorization on checkout, and capture later`.
9. Save your settings.
10. Edit the Cart or Checkout block page.
11. Repeat steps 2–4.
12. Note that the payment method(s) enabled in step 7 are highlighted as not supporting the selected block.

### Actual behavior

When enabling auth + capture, any non-card payment methods enabled beforehand will be identified as not supporting the Card or Checkout blocks.

### Screenshots

CleanShot 2024-08-29 at 20 13 29@2x

Video recording here: U01TTK2RXJB/F07K6NZMPM1-slack-file

### Expected behavior

All payment methods show as compatible with the Cart and Checkout blocks regardless of their state in the WooPayments settings.

### PR DESCRIPTION
**Pull request by mdmoore**
Fixes #9363 

## Changes proposed in this Pull Request

The goal of this change is to prevent payment methods which are not displayed on checkout from causing incompatibility notices in the checkout block. A couple different things lead to this problem:

## Auth + Capture
When manual capture is enabled, LPMs that were already enabled remain enabled even though they're hidden from checkout. This leads to notices about incompatibilities in the checkout block. 

A couple solutions were considered: 

1. Register all payment methods in the checkout block rather than filtering out those that are hidden from checkout. This would require refactoring how we handle adding payment methods via `registerPaymentMethod()`. 
2. Disable the payment methods which are incompatible with manual checkout.

I've went with the second solution. Since these payment methods shouldn't appear on checkout or in the editor, they should be disabled. To do this, I've tapped into the save settings call which updates the enabled payment methods and disables any incompatible payment methods which were previously enabled. The only payment methods compatible with manual capture are `card` and `link`. 

## Inactive payment methods
Payment methods have a capability status, i.e. `active`, `inactive`, `pending`, or `unrequested`. This status isn't taken into account when payment methods are being enabled. All non-`active` status payment methods are hidden from checkout, yet the gateway is still enabled and the payment method id is added to the list of enabled payment methods - `upe_enabled_payment_method_ids`. The solution here is to only enable payment methods that have the `active` status.

### Auth + Capture
1. Enable some LPMs and save the settings.
2. Enable **Issue an authorization on checkout, and capture later** and save the settings.
3. Check that the previously enabled LPMs aren't present in the `upe_enabled_payment_method_ids` option ( easy to view via Dev Tools ). 
4. Open a page with the checkout block and click the Payment Options block. No compatibility notices should appear in the block settings sidebar.

### Inactive payment methods

This requires getting a payment method into a non-active status. The best way I've found to do this is by onboarding a new site. 

1. Create a new JN or similar site. 
2. Onboard WooPayments, selecting General Contractor as the business type.
3. Enable Affirm and save the settings.
4. Affirm should be `rejected` and therefore not added to `upe_enabled_payment_method_ids`. Previously it would've been enabled.
4. Open a page with the checkout block and click the Payment Options block. No compatibility notices should appear in the block settings sidebar.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9470` or branch name `9363-manual-capture-disable-payment-methods` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=9363-manual-capture-disable-payment-methods) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 8f16fac10572c836a96b126501d4869ba89d1936
- Build time: 2024-09-23 21:40:22 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/admin/class-wc-rest-payments-settings-controller.php [hunk 1]
```diff
@@ -603,6 +603,10 @@ private function update_enabled_payment_methods( WP_REST_Request $request ) {
 		$payment_method_ids_to_enable = $request->get_param( 'enabled_payment_method_ids' );
 		$available_payment_methods    = $this->wcpay_gateway->get_upe_available_payment_methods();
 
+		if ( $request->has_param( 'is_manual_capture_enabled' ) && $request->get_param( 'is_manual_capture_enabled' ) ) {
+			$payment_method_ids_to_enable = array_intersect( $payment_method_ids_to_enable, [ 'card', 'link' ] );
```
Comment (danielmx-dev) [inline includes/admin/class-wc-rest-payments-settings-controller.php]: A comment explaining why this is done and why are those payments methods selected would be helpful.

Or maybe just adding the array as a constant in `Payment_Method` with a descriptive name could be enough.

File: includes/migrations/class-manual-capture-payment-method-settings-update.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	 */
+	private function migrate() {
+		$enabled_payment_methods   = $this->main_gateway->get_option( 'upe_enabled_payment_method_ids', [] );
+		$is_manual_capture_enabled = 'yes' === $this->main_gateway->get_option( 'manual_capture' );
+
+		if ( $is_manual_capture_enabled ) {
+			$filtered_payment_methods = array_filter(
+				$enabled_payment_methods,
+				function ( $method ) {
+					return in_array( $method, [ 'card', 'link' ], true );
```
Comment (danielmx-dev) [inline includes/migrations/class-manual-capture-payment-method-settings-update.php]: As mentioned in another file, should a constant be defined for this array ?

File: includes/migrations/class-manual-capture-payment-method-settings-update.php [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+				}
+			);
+
+			foreach ( $this->all_registered_gateways as $gateway ) {
+				$stripe_id = $gateway->get_stripe_id();
+				if ( 'card' !== $stripe_id && 'link' !== $stripe_id ) {
```
Comment (danielmx-dev) [inline includes/migrations/class-manual-capture-payment-method-settings-update.php]: Should the constants be used here?

**Review by danielmx-dev — APPROVED** [review]
Works well, I added a few comments but those shouldn't be blockers 👍

**Comment by mdmoore** [comment]
Thanks for the suggestions @danielmx-dev. Rather adding a new constant just for manual capture payment methods, I've used used the constants already in place and added a comment. Let me know if that sounds alright with you.

**Review by danielmx-dev — APPROVED** [review]
Thanks for the changes, looks good!

**Pull request merged**